pinot-segment-local/src/main/java/org/apache/pinot/segment/local/upsert/BaseTableUpsertMetadataManager.java

```diff
@@ -75,6 +107,127 @@ public void init(TableConfig tableConfig, Schema schema, TableDataManager tableD
 
     _enableSnapshot = upsertConfig.isEnableSnapshot();
     _serverMetrics = serverMetrics;
+    _helixManager = helixManager;
+    _segmentPreloadExecutor = segmentPreloadExecutor;
+    if (_enableSnapshot && upsertConfig.isEnablePreload()) {
+      // Preloading the segments with snapshots for fast upsert metadata recovery.
+      // Note that there is an implicit waiting logic between the thread doing the segment preloading here and the
+      // other helix threads about to process segment state transitions (e.g. taking segments from OFFLINE to ONLINE).
+      // The thread doing the segment preloading here must complete before the other helix threads start to handle
+      // segment state transitions. This is ensured implicitly because segment preloading happens here when
+      // initializing this TableUpsertMetadataManager, which happens when initializing the TableDataManager, which
+      // happens as the lambda of ConcurrentHashMap.computeIfAbsent() method, which ensures the waiting logic.
+      try {
+        _isPreloading = true;
+        preloadSegments();
+      } catch (Exception e) {
+        // Even if preloading fails, we should continue to complete the initialization, so that TableDataManager can be
```

Comment:
This can potentially cause unexpected behavior. Let's not swallow this global catch. If we know certain preload failure can be skipped, we can catch it in a per-segment level so that we can continue preloading the remaining segments

Reply:
good point for finer handling, will update. 

but at a higher level, we'd better not fail the init() method due to preloading failure, so that we can create the tableDataMgr and fall back to the normal loading flow. If init() keeps failing due to uncaught preloading failure, creation of tableDataMgr would be retried, leading to another round of preloading while calling the init() method again.